app/javascript/mastodon/components/column_header.tsx

```diff
@@ -0,0 +1,309 @@
+
+import type { IconProp } from 'mastodon/components/icon';
+import { Icon } from 'mastodon/components/icon';
+import {
+  ButtonInTabsBar,
+  useColumnsContext,
+} from 'mastodon/features/ui/util/columns_context';
+import type { ContextWithIdentity } from 'mastodon/utils/identity';
+
+import { useAppHistory } from './router';
+
+const messages = defineMessages({
+  show: { id: 'column_header.show_settings', defaultMessage: 'Show settings' },
+  hide: { id: 'column_header.hide_settings', defaultMessage: 'Hide settings' },
+  moveLeft: {
+    id: 'column_header.moveLeft_settings',
+    defaultMessage: 'Move column to the left',
+  },
+  moveRight: {
+    id: 'column_header.moveRight_settings',
+    defaultMessage: 'Move column to the right',
+  },
+});
+
+interface BackButtonProps {
+  pinned?: boolean;
+  show?: boolean;
+}
+
+const BackButton = ({ pinned, show }: BackButtonProps) => {
+  const history = useAppHistory();
+  const { multiColumn } = useColumnsContext();
+
+  const handleBackClick = useCallback(() => {
+    if (history.location.state?.fromMastodon) {
+      history.goBack();
+    } else {
+      history.push('/');
+    }
+  }, [history]);
+
+  const showButton =
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    history &&
+    !pinned &&
+    ((multiColumn && history.location.state?.fromMastodon) || show);
+
+  if (!showButton) return null;
+
+  return (
+    <button onClick={handleBackClick} className='column-header__back-button'>
+      <Icon
+        id='chevron-left'
+        icon={ArrowBackIcon}
+        className='column-back-button__icon'
+      />
+      <FormattedMessage id='column_back_button.label' defaultMessage='Back' />
+    </button>
+  );
+};
+
+interface ColumnHeaderProps extends RouteComponentProps {
+  active?: boolean;
+  appendContent?: React.ReactNode;
+  children?: React.ReactNode;
+  collapseIssues?: boolean;
+  extraButton?: React.ReactNode;
+  icon?: string;
+  iconComponent?: IconProp;
+  intl: IntlShape;
+  multiColumn?: boolean;
+  onClick?: () => void;
+  onMove?: (direction: number) => void;
+  onPin?: () => void;
+  pinned?: boolean;
+  placeholder?: boolean;
+  showBackButton?: boolean;
+  title?: React.ReactNode;
+}
+
+class ColumnHeaderInternal extends PureComponent<ColumnHeaderProps> {
+  declare context: ContextWithIdentity;
+
+  static contextTypes = {
+    identity: PropTypes.object,
+  };
+
+  state = {
+    collapsed: true,
+    animating: false,
+  };
+
+  handleToggleClick = (e: React.MouseEvent) => {
+    e.stopPropagation();
+    this.setState({ collapsed: !this.state.collapsed, animating: true });
+  };
+
+  handleTitleClick = () => {
+    this.props.onClick?.();
+  };
+
+  handleMoveLeft = () => {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    this.props.onMove!(-1);
+  };
+
+  handleMoveRight = () => {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```

Comment:
I wasn't sure exactly what the prop shape was here: is it that `onMove` is required _if_ `multiColumn` is true?